pkgs/development/haskell-modules/make-package-set.nix

```diff
@@ -257,20 +261,46 @@ in package-set { inherit pkgs stdenv callPackage; } self // {
     shellFor = { packages, withHoogle ? false, ... } @ args:
       let
         selected = packages self;
-        packageInputs = builtins.map getBuildInputs selected;
-        haskellInputs =
-          builtins.filter
-            (input: pkgs.lib.all (p: input.outPath != p.outPath) selected)
-            (pkgs.lib.concatMap (p: p.haskellBuildInputs) packageInputs);
+
+        packageInputs = map getBuildInputs selected;
+
+        # If `packages = [ a b ]` and `a` depends on `b`, don't build `b`, such
+        # that `a` will take the `b` dependency from the cabal.project file.
+        # This allows changing `b` while in the shell and have `a` use the
+        # changes directly.
+        haskellInputs = pkgs.lib.filter
+          (input: pkgs.lib.all (p: input.outPath != p.outPath) selected)
+          (pkgs.lib.concatMap (p: p.haskellBuildInputs) packageInputs);
         systemInputs = pkgs.lib.concatMap (p: p.systemBuildInputs) packageInputs;
+
         withPackages = if withHoogle then self.ghcWithHoogle else self.ghcWithPackages;
+        ghcEnv = withPackages (p: haskellInputs);
+        nativeBuildInputs = pkgs.lib.concatMap (p: p.nativeBuildInputs) selected;
+
+        ghcCommand' = if ghc.isGhcjs or false then "ghcjs" else "ghc";
+        ghcCommand = "${ghc.targetPrefix}${ghcCommand'}";
+        ghcCommandCaps= pkgs.lib.toUpper ghcCommand';
+
         mkDrvArgs = builtins.removeAttrs args ["packages" "withHoogle"];
       in pkgs.stdenv.mkDerivation (mkDrvArgs // {
-        name = "ghc-shell-for-packages";
-        nativeBuildInputs = [(withPackages (_: haskellInputs))] ++ mkDrvArgs.nativeBuildInputs or [];
+        name = "interactive-${pkgs.lib.concatMapStringsSep "-" (p: p.name) selected}-environment";
+
         buildInputs = systemInputs ++ mkDrvArgs.buildInputs or [];
+        nativeBuildInputs = [ ghcEnv ] ++ nativeBuildInputs ++ mkDrvArgs.nativeBuildInputs or [];
         phases = ["installPhase"];
         installPhase = "echo $nativeBuildInputs $buildInputs > $out";
+        LANG = "en_US.UTF-8";
+        LOCALE_ARCHIVE = pkgs.lib.optionalString (stdenv.hostPlatform.libc == "glibc") "${buildPackages.glibcLocales}/lib/locale/locale-archive";
```

Comment:
Should these be here? I think they're only desirable with `nix-shell --pure`?

Reply:
They've always been in the `.env`. I don't think this has a negative impact.

Reply:
It does if anyone likes the LANG on their system :P

Reply:
Is `--pure` detectable though?

If it's not, I'd say we should keep it. I once had it that `readFile :: FilePath -> Text` failed with a decoding error because it was a pure shell and didn't have these vars set (or was it `LC_ALL`?, not sure).

And to be backwards compatible.

Reply:
It'd be pretty bad to try to detect `--pure` anyway, I guess. So yea, this should stay.